exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,398 @@
+    /// - A transaction execution status (see `ExecutionResult` and `TransactionResult` for the
+    ///   details) is stored in the blockchain and can be accessed through api.
+    /// - Blockchain state hash is affected by the transactions execution status.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::{Transaction, ExecutionResult};
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #
+    /// #       public_key: &PublicKey,
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Return execution status.
+    ///         Ok(())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult;
+}
+
+/// Result of unsuccessful transaction execution.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExecutionError {
+    /// General failure (unspecified reason).
+    Failure,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+    /// User-defined string error description.
+    Description(String),
+}
+
+/// Extended by the framework result of unsuccessful transaction execution.
+///
+/// # Notes:
+///
+/// - `Description`'s content excluded from hash calculation (see `StorageValue` implementation for
+///   the details).
+/// - `TransactionError::Panic` is set by the framework if panic is raised during transaction
+///   execution.
+/// - `TransactionError` implements `Display` which can be used for obtaining a simple error
+///   description.
+///
+/// # Examples
+///
+/// ```
+/// use exonum::blockchain::TransactionError;
+///
+/// # let transaction_error = TransactionError::Panic;
+/// // Prints user friendly error description.
+/// println!("Transaction error: {}", transaction_error);
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionError {
+    /// Panic occurred during transaction execution.
+    Panic,
+    /// General failure (unspecified reason).
+    Failure,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+    /// User-defined string error description.
+    Description(String),
+}
+
+/// Converts from `ExecutionResult` into `TransactionResult`.
+pub fn convert_result(result: ExecutionResult) -> TransactionResult {
```

Comment:
Why isn't this method implemented as `impl From<ExecutionResult> for TransactionResult`?

Reply:
Two reasons:
1. I don't think it should be part of the public API: user should use both types separately without converting one to another.
2. Rust [doesn't allow this](https://play.rust-lang.org/?gist=3c7bdc53ebd157379316c546b677fd1d&version=stable). :smile:

Reply:
Terribly sorry; I meant using `impl From<ExecutionError> for TransactionError` + `map_err(TransactionError::from)`.

Reply:
Still I don't like extending public API with such conversion: will it be useful for exonum users?.. On the one hand, it seems harmless, on the other hand, it provides more room for breaking changes.

Reply:
I think it is an intended addition to the `ExecutionError` / `TransactionError` interface: it seems an invariant w.r.t. future changes that `ExecutionError`s are a subset of `TransactionError`s. But I don't insist on this change.

Reply:
Makes sense.